python/ray/serve/controller.py

```diff
@@ -276,55 +487,51 @@ async def _recover_from_checkpoint(self, checkpoint_bytes: bytes) -> None:
         start = time.time()
         logger.info("Recovering from checkpoint")
 
-        # Load internal state from the checkpoint data.
-        (
-            self.routes,
-            router_node_ids,
-            self.backends,
-            self.traffic_policies,
-            self.replicas,
-            self.replicas_to_start,
-            self.replicas_to_stop,
-            self.backends_to_remove,
-            self.endpoints_to_remove,
-        ) = pickle.loads(checkpoint_bytes)
-
-        for node_id in router_node_ids:
+        restored_checkpoint: Checkpoint = pickle.loads(checkpoint_bytes)
+        # Restore ConfigurationStore
+        self.configuration_store = restored_checkpoint.config
+
+        # Restore ActorNursery
+        self.actor_nursery = restored_checkpoint.nursery
+
+        # Refresh the RouterCache
+        for node_id in self.actor_nursery.routers_cache.keys():
             router_name = format_actor_name(SERVE_PROXY_NAME,
                                             self.controller_name, node_id)
-            self.routers[node_id] = ray.get_actor(router_name)
-
+            self.actor_nursery.routers_cache[node_id] = ray.get_actor(
+                router_name)
         # Fetch actor handles for all of the backend replicas in the system.
         # All of these workers are guaranteed to already exist because they
```

Comment:
It might make sense to move this to actor_nursery like `actor_nursery.restore_from_checkpoint`

Reply:
even a custom deserializer?